File: ureka-framework-main/ureka_framework/logic/pipeline_flow/flow_issue_u_ticket.py
# Deployment Environment
from ureka_framework.environment import Environment

# Data Model (RAM)
from ureka_framework.model.shared_data import SharedData
import ureka_framework.model.data_model.this_device as this_device

# Data Model (Message)
import ureka_framework.model.message_model.message as message
import ureka_framework.model.message_model.u_ticket as u_ticket
from ureka_framework.model.message_model.u_ticket import UTicket, jsonstr_to_u_ticket
import ureka_framework.model.message_model.r_ticket as r_ticket
from ureka_framework.model.message_model.r_ticket import RTicket

# Resource (Logger)
from ureka_framework.resource.logger.simple_logger import simple_log

# Measure Helper
from ureka_framework.logic.stage_worker.measure_helper import MeasureHelper

# Stage Worker
from ureka_framework.logic.stage_worker.received_msg_storer import ReceivedMsgStorer
from ureka_framework.logic.stage_worker.msg_verifier import MsgVerifier
from ureka_framework.logic.stage_worker.executor import Executor
from ureka_framework.logic.stage_worker.msg_generator import MsgGenerator
from ureka_framework.logic.stage_worker.generated_msg_storer import GeneratedMsgStorer
from ureka_framework.logic.stage_worker.msg_sender import MsgSender


class FlowIssueUTicket:
    def __init__(
        self,
        share_data: SharedData,
        measure_helper: MeasureHelper,
        received_msg_storer: ReceivedMsgStorer,
        msg_verifier: MsgVerifier,
        executor: Executor,
        msg_generator: MsgGenerator,
        generated_msg_storer: GeneratedMsgStorer,
        msg_sender: MsgSender,
    ) -> None:
        self.shared_data = share_data
        self.measure_helper = measure_helper
        self.received_msg_storer = received_msg_storer
        self.msg_verifier = msg_verifier
        self.executor = executor
        self.msg_generator = msg_generator
        self.generated_msg_storer = generated_msg_storer
        self.msg_sender = msg_sender

    ######################################################
    # [PIPELINE FLOW]
    #
    # CST: issuer_issue_u_ticket_to_herself()
    # TODO: REQ: _issuer_recv_request() <- holder_send_request_to_issuer()
    # CST: issuer_issue_u_ticket_to_holder() -> _holder_recv_u_ticket()
    # RTN: _issuer_recv_r_ticket() <- holder_send_r_ticket_to_issuer()
    #
    # TODO: More complete Tx (with DID, etc.))
    # TODO: Rollback (e.g., delete the temporary stored state and stored message) if fail
    #         execution only change state after success, but need pay attention to (SR)
    # TODO: QoS Testing - Re-transmission, & Timeout
    #       (e.g., if RTicket is not returned, holder can request backup RTicket)
    #       (e.g., if CR or PS is Timeout, device can revert to the WAIT_FOR_UT state)
    ######################################################
    def issuer_issue_u_ticket_to_herself(
        self, device_id: str, arbitrary_dict: dict
    ) -> None:
        ######################################################
        # Start Process Measurement
        ######################################################
        self.measure_helper.measure_process_perf_start()

        try:
            # [STAGE: (VL)]
            if device_id in self.shared_data.device_table or device_id == "no_id":
                # [STAGE: (G)]
                generated_u_ticket_json: str = (
                    self.msg_generator._generate_xxx_u_ticket(arbitrary_dict)
                )
                # simple_log("debug", f"Generated UTicket: {generated_u_ticket_json}")

                # [STAGE: (SG)]
                self.generated_msg_storer._store_generated_xxx_u_ticket(
                    generated_u_ticket_json
                )

                # [STAGE: (O)]
                self.executor._execute_update_ticket_order(
                    "holder-generate-or-receive-uticket",
                    jsonstr_to_u_ticket(generated_u_ticket_json),
                )

        except (
            RuntimeError
        ):  # pragma: no cover -> Weird Ticket-Request (ValidationError)
            failure_msg = f"FAILURE: (VUREQ)"
            simple_log("error", failure_msg)

        except:  # pragma: no cover -> Shouldn't Reach Here
            raise RuntimeError(f"Shouldn't Reach Here")

        finally:
            ######################################################
            # End Process Measurement
            ######################################################
            self.measure_helper.measure_recv_cli_perf_time(
                "issuer_issue_u_ticket_to_herself"
            )

            ########################################################################
            # Start Comm Measurement
            ########################################################################
            if self.shared_data.this_device.device_name != "iot_device":
                self.measure_helper.measure_comm_perf_start()

    def issuer_issue_u_ticket_to_holder(
        self, device_id: str, arbitrary_dict: dict
    ) -> None:
        ######################################################
        # Start Process Measurement
        ######################################################
        self.measure_helper.measure_process_perf_start()

        try:
            # [STAGE: (VL)]
            if device_id in self.shared_data.device_table:
                # [STAGE: (G)]
                generated_u_ticket_json: str = (
                    self.msg_generator._generate_xxx_u_ticket(arbitrary_dict)
                )
                # simple_log("debug", f"Generated UTicket: {generated_u_ticket_json}")

                # [STAGE: (SG)]
                self.generated_msg_storer._store_generated_xxx_u_ticket(
                    generated_u_ticket_json
                )

                ######################################################
                # End Process Measurement
                ######################################################
                self.measure_helper.measure_recv_cli_perf_time(
                    "issuer_issue_u_ticket_to_holder"
                )

                ########################################################################
                # Start Comm Measurement
                ########################################################################
                if self.shared_data.this_device.device_name != "iot_device":
                    self.measure_helper.measure_comm_perf_start()

                # [STAGE: (S)]
                self.msg_sender._send_xxx_message(
                    message.MESSAGE_RECV_AND_STORE,
                    u_ticket.MESSAGE_TYPE,
                    generated_u_ticket_json,
                )

        except KeyError:  # pragma: no cover -> FAILURE: (VL)
            error = "FAILURE: (VL)"
            self.shared_data.result_message = f"{error}"
            raise RuntimeError(f"{error}")

        except (
            RuntimeError
        ):  # pragma: no cover -> Weird Ticket-Request (ValidationError)
            failure_msg = f"FAILURE: (VUREQ)"
            simple_log("error", failure_msg)

        except:  # pragma: no cover -> Shouldn't Reach Here
            raise RuntimeError(f"Shouldn't Reach Here")

        # Manually Finish Simulated Comm
        simple_log(
            "debug",
            f"+ {self.shared_data.this_device.device_name} manually finish UT-UT~~ (issuer)",
        )
        if Environment.COMMUNICATION_CHANNEL == "SIMULATED":
            self.msg_sender.complete_simulated_comm()

    def _holder_recv_u_ticket(self, received_u_ticket: UTicket) -> None:
        try:
            # [STAGE: (R)(VR)]
            # But the actual device id & ticket order in device is still unknown
            # -> TODO: test_fail_when_double_issuing_or_double_spending

            # [STAGE: (SR)]
            self.received_msg_storer._store_received_xxx_u_ticket(received_u_ticket)

            # [STAGE: (O)]
            self.executor._execute_update_ticket_order(
                "holder-generate-or-receive-uticket", received_u_ticket
            )

        except:  # pragma: no cover -> Shouldn't Reach Here
            raise RuntimeError(f"Shouldn't Reach Here")

        finally:
            # [STAGE: (G)(S)]
            # Can optionally _generate_xxx_r_ticket & _send_xxx_message

            pass

            ######################################################
            # End Process Measurement
            ######################################################
            self.measure_helper.measure_recv_msg_perf_time(
                "_holder_recv_u_ticket",
            )

            # ########################################################################
            # # Start Comm Measurement
            # ########################################################################
            # if self.shared_data.this_device.device_name != "iot_device":
            #     self.measure_helper.measure_comm_perf_start()

        # Manually Finish Simulated Comm
        simple_log(
            "debug",
            f"+ {self.shared_data.this_device.device_name} manually finish UT-UT~~ (holder)",
        )
        if Environment.COMMUNICATION_CHANNEL == "SIMULATED":
            self.msg_sender.complete_simulated_comm()

    def holder_send_r_ticket_to_issuer(self, device_id: str) -> None:
        ######################################################
        # Start Process Measurement
        ######################################################
        self.measure_helper.measure_process_perf_start()

        try:
            # [STAGE: (VL)(L)]
            stored_r_ticket_json: str = self.shared_data.device_table[
                device_id
            ].device_r_ticket_for_owner
            # simple_log("debug", f"Stored RTicket: {stored_r_ticket_json}")

            ######################################################
            # End Process Measurement
            ######################################################
            self.measure_helper.measure_recv_cli_perf_time(
                "holder_send_r_ticket_to_issuer"
            )

            ########################################################################
            # Start Comm Measurement
            ########################################################################
            if self.shared_data.this_device.device_name != "iot_device":
                self.measure_helper.measure_comm_perf_start()

            # [STAGE: (S)]
            self.msg_sender._send_xxx_message(
                message.MESSAGE_RECV_AND_STORE,
                r_ticket.MESSAGE_TYPE,
                stored_r_ticket_json,
            )

        except KeyError:  # pragma: no cover -> FAILURE: (VL)
            error = "FAILURE: (VL)"
            self.shared_data.result_message = f"{error}"
            raise RuntimeError(f"{error}")

        except:  # pragma: no cover -> Shouldn't Reach Here
            raise RuntimeError(f"Shouldn't Reach Here")

        # Manually Finish Simulated Comm
        simple_log(
            "debug",
            f"+ {self.shared_data.this_device.device_name} manually finish RT-RT~~ (holder)",
        )
        if Environment.COMMUNICATION_CHANNEL == "SIMULATED":
            self.msg_sender.complete_simulated_comm()

    def _issuer_recv_r_ticket(self, received_r_ticket: RTicket) -> None:
        try:
            # [STAGE: (R)(VR)]

            # [STAGE: (SR)]
            self.received_msg_storer._store_received_xxx_r_ticket(received_r_ticket)

            if (
                received_r_ticket.r_ticket_type == u_ticket.TYPE_INITIALIZATION_UTICKET
                or received_r_ticket.r_ticket_type == u_ticket.TYPE_OWNERSHIP_UTICKET
                or received_r_ticket.r_ticket_type == u_ticket.TYPE_ACCESS_END_UTOKEN
            ):
                # Query Corresponding UTicket(s)
                #   Notice that even Initialization UTicket is copied in the device_table["device_id"]
                # [STAGE: (VL)(L)]
                if received_r_ticket.r_ticket_type == u_ticket.TYPE_OWNERSHIP_UTICKET:
                    stored_u_ticket_json: str = self.shared_data.device_table[
                        received_r_ticket.device_id
                    ].device_ownership_u_ticket_for_others
                elif received_r_ticket.r_ticket_type == u_ticket.TYPE_ACCESS_END_UTOKEN:
                    stored_u_ticket_json: str = self.shared_data.device_table[
                        received_r_ticket.device_id
                    ].device_access_u_ticket_for_others
                else:  # pragma: no cover -> Shouldn't Reach Here
                    raise RuntimeError(f"Shouldn't Reach Here")
                simple_log("debug", f"Corresponding UTicket: {stored_u_ticket_json}")
                # [STAGE: (VR)]
                stored_u_ticket = self.msg_verifier._classify_u_ticket_is_defined_type(
                    stored_u_ticket_json
                )

                # [STAGE: (VRT)]
                self.msg_verifier.verify_u_ticket_has_executed_through_r_ticket(
                    r_ticket_in=received_r_ticket,
                    audit_start_ticket=stored_u_ticket,
                    audit_end_ticket=None,
                )

                # [STAGE: (E)(O)]
                self.executor._execute_xxx_r_ticket(
                    r_ticket_in=received_r_ticket, comm_end="issuer"
                )
                self.shared_data.result_message = f"-> SUCCESS: VERIFY_UT_HAS_EXECUTED"

                # [STAGE: (C)]
                self.executor._change_state(
                    this_device.STATE_AGENT_WAIT_FOR_UREQ_UREJ_UT_RT
                )

            else:  # pragma: no cover -> TODO: Revocation UTicket
                # Query Corresponding UTicket(s)
                raise RuntimeError(f"Not implemented yet")

        except KeyError:  # pragma: no cover -> FAILURE: (VL)
            error = "FAILURE: (VL)"
            self.shared_data.result_message = f"{error}"
            raise RuntimeError(f"{error}")

        except RuntimeError as error:  # pragma: no cover -> FAILURE: (VR)(VRT)
            # TODO: (VRT) test_fail_when_double_issuing_or_double_spending
            error = "FAILURE: (VR)(VRT)"
            self.shared_data.result_message = f"{error}"
            raise RuntimeError(f"{error}")

        except:  # pragma: no cover -> Shouldn't Reach Here
            raise RuntimeError(f"Shouldn't Reach Here")

        finally:
            simple_log("debug", f"result_message = {self.shared_data.result_message}")

            # ######################################################
            # # End Process Measurement
            # ######################################################
            # self.measure_helper.measure_recv_msg_perf_time(
            #     "_issuer_recv_r_ticket",
            # )

            # ########################################################################
            # # Start Comm Measurement
            # ########################################################################
            # if self.shared_data.this_device.device_name != "iot_device":
            #     self.measure_helper.measure_comm_perf_start()

        # Manually Finish Simulated Comm
        simple_log(
            "debug",
            f"+ {self.shared_data.this_device.device_name} manually finish RT-RT~~ (issuer)",
        )
        if Environment.COMMUNICATION_CHANNEL == "SIMULATED":
            self.msg_sender.complete_simulated_comm()

File: UrekaApp/app/src/main/java/ureka/framework/logic/pipeline_flow/FlowIssueUTicket.java
package ureka.framework.logic.pipeline_flow;

import com.example.urekaapp.AdminAgentActivity;
import com.example.urekaapp.VoterAgentActivity;

import java.security.KeyException;
import java.util.Map;

import ureka.framework.Environment;
import ureka.framework.logic.stage_worker.Executor;
import ureka.framework.logic.stage_worker.GeneratedMsgStorer;
import ureka.framework.logic.stage_worker.MeasureHelper;
import ureka.framework.logic.stage_worker.MsgGenerator;
import ureka.framework.logic.stage_worker.MsgSender;
import ureka.framework.logic.stage_worker.MsgVerifier;
import ureka.framework.logic.stage_worker.ReceivedMsgStorer;
import ureka.framework.model.SharedData;
import ureka.framework.model.data_model.OtherDevice;
import ureka.framework.model.data_model.ThisDevice;
import ureka.framework.model.message_model.Message;
import ureka.framework.model.message_model.RTicket;
import ureka.framework.model.message_model.UTicket;
import ureka.framework.resource.logger.SimpleLogger;

public class FlowIssueUTicket {
    private SharedData sharedData;
    private MeasureHelper measureHelper;
    private ReceivedMsgStorer receivedMsgStorer;
    private MsgVerifier msgVerifier;
    private Executor executor;
    private MsgGenerator msgGenerator;
    private GeneratedMsgStorer generatedMsgStorer;
    private MsgSender msgSender;

    public FlowIssueUTicket(SharedData sharedData, MeasureHelper measureHelper,
                            ReceivedMsgStorer receivedMsgStorer, MsgVerifier msgVerifier,
                            Executor executor, MsgGenerator msgGenerator,
                            GeneratedMsgStorer generatedMsgStorer, MsgSender msgSender) {
        this.sharedData = sharedData;
        this.measureHelper = measureHelper;
        this.receivedMsgStorer = receivedMsgStorer;
        this.msgVerifier = msgVerifier;
        this.executor = executor;
        this.msgGenerator = msgGenerator;
        this.generatedMsgStorer = generatedMsgStorer;
        this.msgSender = msgSender;
    }

    public void setSharedData(SharedData sharedData) {
        this.sharedData = sharedData;
    }

    public SharedData getSharedData() {
        return sharedData;
    }

    public void setMeasureHelper(MeasureHelper measureHelper) {
        this.measureHelper = measureHelper;
    }

    public MeasureHelper getMeasureHelper() {
        return measureHelper;
    }

    public void setReceivedMsgStorer(ReceivedMsgStorer receivedMsgStorer) {
        this.receivedMsgStorer = receivedMsgStorer;
    }

    public ReceivedMsgStorer getReceivedMsgStorer() {
        return receivedMsgStorer;
    }

    public void setMsgVerifier(MsgVerifier msgVerifier) {
        this.msgVerifier = msgVerifier;
    }

    public MsgVerifier getMsgVerifier() {
        return msgVerifier;
    }

    public void setExecutor(Executor executor) {
        this.executor = executor;
    }

    public Executor getExecutor() {
        return executor;
    }

    public void setMsgGenerator(MsgGenerator msgGenerator) {
        this.msgGenerator = msgGenerator;
    }

    public MsgGenerator getMsgGenerator() {
        return msgGenerator;
    }

    public void setGeneratedMsgStorer(GeneratedMsgStorer generatedMsgStorer) {
        this.generatedMsgStorer = generatedMsgStorer;
    }

    public GeneratedMsgStorer getGeneratedMsgStorer() {
        return generatedMsgStorer;
    }

    public void setMsgSender(MsgSender msgSender) {
        this.msgSender = msgSender;
    }

    public MsgSender getMsgSender() {
        return msgSender;
    }
    /* [PIPELINE FLOW]
        CST: issuer_issue_u_ticket_to_herself()
        TODO: REQ: _issuer_recv_request() <- holder_send_request_to_issuer()
        CST: issuer_issue_u_ticket_to_holder() -> _holder_recv_u_ticket()
        RTN: _issuer_recv_r_ticket() <- holder_send_r_ticket_to_issuer()

        TODO: More complete Tx (with DID, etc.))
        TODO: Rollback (e.g., delete the temporary stored state and stored message) if fail
             execution only change state after success, but need pay attention to (SR)
        TODO: QoS Testing - Re-transmission, & Timeout
            (e.g., if RTicket is not returned, holder can request backup RTicket)
            (e.g., if CR or PS is Timeout, device can revert to the WAIT_FOR_UT state)
    */

    public void issuerIssueUTicketToHerself(String device_id, Map<String, String> arbitraryDict) {
        // Start Process Measurement
        this.measureHelper.measureProcessPerfStart();
        try {
            // [STAGE: (VL)]
            SimpleLogger.simpleLog("info", "deviceId = " + device_id + ", containing = " + (this.sharedData.getDeviceTable().containsKey(device_id) || "no_id".equals(device_id)));
            if (this.sharedData.getDeviceTable().containsKey(device_id) || "no_id".equals(device_id)) {
                // [STAGE: (G)]
                String generatedUTicketJson = this.msgGenerator.generateXxxUTicket(arbitraryDict);

                // [STAGE: (SG)]
                this.generatedMsgStorer.storeGeneratedXxxUTicket(generatedUTicketJson);

                // [STAGE: (O)]
                this.executor.executeUpdateTicketOrder("holderGenerateOrReceiveUTicket",UTicket.jsonStrToUTicket(generatedUTicketJson));
            }
        } catch (RuntimeException e) { // pragma: no cover -> Weird Ticket-Request (ValidationError)
            SimpleLogger.simpleLog("error", "FAILURE: (VUREQ)");
        } catch (Exception e) { // pragma: no cover -> Shouldn't Reach Here
            throw new RuntimeException("Shouldn't Reach Here", e);
        } finally {
            // End Process Measurement
            this.measureHelper.measureRecvCliPerfTime("issuerIssueUTicketToHerself");

            // Start Comm Measurement
            if (!this.sharedData.getThisDevice().getDeviceName().equals("iotDevice")) {
                this.measureHelper.measureCommPerfStart();
            }
        }
    }

    public void issuerIssueUTicketToHolder(Map<String, String> arbitraryDict) {
        // Start Process Measurement
        this.measureHelper.measureProcessPerfStart();
        try {
            // [STAGE: (VL)]
            if (this.sharedData.getDeviceTable().containsKey(AdminAgentActivity.connectedDeviceId)) {
                // [STAGE: (G)]
                arbitraryDict.put("device_id", AdminAgentActivity.connectedDeviceId);
                String generatedUTicketJson = this.msgGenerator.generateXxxUTicket(arbitraryDict);
//                UTicket generatedUTicket = UTicket.jsonStrToUTicket(generatedUTicketJson);
//                generatedUTicket.setDeviceId(AdminAgentActivity.connectedDeviceId);
//                generatedUTicketJson = UTicket.uTicketToJsonStr(generatedUTicket);

                // [STAGE: (SG)]
                this.generatedMsgStorer.storeGeneratedXxxUTicket(generatedUTicketJson);

                // End Process Measurement
                this.measureHelper.measureRecvCliPerfTime("issuerIssueUTicketToHolder");

                // Start Comm Measurement
                if (!this.sharedData.getThisDevice().getDeviceName().equals("iotDevice")) {
                    this.measureHelper.measureCommPerfStart();
                }

                // [STAGE: (S)]
                this.msgSender.sendXxxMessageByNearby(
                        Message.MESSAGE_RECV_AND_STORE,
                        UTicket.MESSAGE_TYPE,
                        generatedUTicketJson
                );
            } else {
                SimpleLogger.simpleLog("info", "FlowIssueUTicket.issuerIssueUTicketToHolder: Device not in device table");
            }
        } catch (RuntimeException e) { // pragma: no cover -> Weird Ticket-Request (ValidationError)
            SimpleLogger.simpleLog("error", "FAILURE: (VUREQ)");
        } catch (Exception e) { // pragma: no cover -> Shouldn't Reach Here
            throw new RuntimeException("Shouldn't Reach Here", e);
        }
        // Manually Finish Simulated Comm
        SimpleLogger.simpleLog("debug", this.sharedData.getThisDevice().getDeviceName() + " manually finish UT-UT~~ (issuer)");
        if (Environment.COMMUNICATION_CHANNEL.equals("SIMULATED")) {
            this.msgSender.completeSimulatedComm();
        }
    }

    public void _holderRecvUTicket(UTicket receivedUTicket) {
        try {
            /*
             [STAGE: (R)(VR)]
             But the actual device id & ticket order in device is still unknown
             -> TODO: test_fail_when_double_issuing_or_double_spending
            */

            // [STAGE: (SR)]
            VoterAgentActivity.connectedDeviceId = receivedUTicket.getDeviceId();
            this.receivedMsgStorer.storeReceivedXxxUTicket(receivedUTicket);

            // [STAGE: (O)]
            this.executor.executeUpdateTicketOrder("holderGenerateOrReceiveUTicket", receivedUTicket);
        } catch (Exception e) { // pragma: no cover -> Shouldn't Reach Here
            throw new RuntimeException("Shouldn't Reach Here", e);
        } finally {
            // [STAGE: (G)(S)]
            // Can optionally _generateXxxRTicket & _sendXxxMessage

            // End Process Measurement
            this.measureHelper.measureRecvMsgPerfTime("_holderRecvUTicket");
        }
        // Manually Finish Simulated Comm
        SimpleLogger.simpleLog("debug", sharedData.getThisDevice().getDeviceName() + " manually finish UT-UT~~ (holder)");
        if (Environment.COMMUNICATION_CHANNEL.equals("SIMULATED")) {
            msgSender.completeSimulatedComm();
        }
    }

    public void holderSendRTicketToIssuer(String device_id) {
        // Start Process Measurement
        this.measureHelper.measureProcessPerfStart();
        try {
            // [STAGE: (VL)(L)]
            String storedRTicketJson = this.sharedData.getDeviceTable().get(device_id).getDeviceRTicketForOwner();

            // End Process Measurement
            this.measureHelper.measureRecvCliPerfTime("holderSendRTicketToIssuer");

            // Start Comm Measurement
            if (!this.sharedData.getThisDevice().getDeviceName().equals("iotDevice")) {
                this.measureHelper.measureCommPerfStart();
            }

            // [STAGE: (S)]
            this.msgSender.sendXxxMessage(
                    Message.MESSAGE_RECV_AND_STORE,
                    RTicket.MESSAGE_TYPE,
                    storedRTicketJson
            );
        } catch (Exception e) { // pragma: no cover -> Shouldn't Reach Here
            throw new RuntimeException("Shouldn't Reach Here", e);
        }
        // Manually Finish Simulated Comm
        SimpleLogger.simpleLog("debug", this.sharedData.getThisDevice().getDeviceName() + " manually finish RT-RT~~ (holder)");
        if (Environment.COMMUNICATION_CHANNEL.equals("SIMULATED")) {
            this.msgSender.completeSimulatedComm();
        }
    }

    public void _issuerRecvRTicket(RTicket receivedRTicket) {
        try {
            // [STAGE: (R)(VR)]
            // [STAGE: (SR)]
            this.receivedMsgStorer.storeReceivedXxxRTicket(receivedRTicket);
            String storedUTicketJson;
            if(receivedRTicket.getRTicketType().equals(UTicket.TYPE_INITIALIZATION_UTICKET) ||
                    receivedRTicket.getRTicketType().equals(UTicket.TYPE_OWNERSHIP_UTICKET) ||
                    receivedRTicket.getRTicketType().equals(UTicket.TYPE_ACCESS_END_UTOKEN)) {
                // Query Corresponding UTicket(s)
                //    Notice that even Initialization UTicket is copied in the deviceTable["device_id"]
                // [STAGE: (VL)(L)]
                if(receivedRTicket.getRTicketType().equals(UTicket.TYPE_OWNERSHIP_UTICKET)) {
                    storedUTicketJson = this.sharedData.getDeviceTable().get(receivedRTicket.getDeviceId()).getDeviceOwnershipUTicketForOthers();
                } else if (receivedRTicket.getRTicketType().equals(UTicket.TYPE_ACCESS_END_UTOKEN)) {
                    storedUTicketJson = sharedData.getDeviceTable().get(receivedRTicket.getDeviceId()).getDeviceAccessUTicketForOthers();
                } else {
                    throw new RuntimeException("Shouldn't Reach Here");
                }

                SimpleLogger.simpleLog("debug", "Corresponding UTicket: " + storedUTicketJson);

                // [STAGE: (VR)]
                UTicket storedUTicket = this.msgVerifier._classifyUTicketIsDefinedType(storedUTicketJson);

                // [STAGE: (VRT)]
                this.msgVerifier.verifyUTicketHasExecutedThroughRTicket(receivedRTicket, storedUTicket, null);

                // [STAGE: (E)(O)]
                this.executor.executeXxxRTicket(receivedRTicket, "issuer");
                this.sharedData.setResultMessage("-> SUCCESS: VERIFY_UT_HAS_EXECUTED");

                // [STAGE: (C)]
                this.executor.changeState(ThisDevice.STATE_AGENT_WAIT_FOR_UREQ_UREJ_UT_RT);
            } else { // pragma: no cover -> TODO: Revocation UTicket
                throw new RuntimeException("Not implemented yet");
            }
        } catch (RuntimeException e) { // pragma: no cover -> FAILURE: (VR)(VRT)
            // TODO: (VRT) test_fail_when_double_issuing_or_double_spending
            this.sharedData.setResultMessage("FAILURE: (VR)(VRT)");
            throw new RuntimeException("FAILURE: (VR)(VRT)", e);
        } catch (Exception e) { // pragma: no cover -> Shouldn't Reach Here
            throw new RuntimeException("Shouldn't Reach Here", e);
        } finally {
            SimpleLogger.simpleLog("debug", "result_message = " + this.sharedData.getResultMessage());
        }

        SimpleLogger.simpleLog("debug", this.sharedData.getThisDevice().getDeviceName() + " manually finish RT-RT~~ (issuer)");
        if (Environment.COMMUNICATION_CHANNEL.equals("SIMULATED")) {
            this.msgSender.completeSimulatedComm();
        }
    }
}